modules/dspxBidAdapter.js

```diff
@@ -296,4 +388,120 @@ function getMediaTypesInfo(bid) {
   return mediaTypesInfo;
 }
 
+/**
+ * Get Bid Floor
+ * @param bid
+ * @returns {number|*}
+ */
+function getBidFloor(bid) {
+  if (!isFn(bid.getFloor)) {
+    return deepAccess(bid, 'params.bidfloor', 0);
+  }
+
+  try {
+    const bidFloor = bid.getFloor({
+      currency: 'EUR',
+      mediaType: '*',
+      size: '*',
+    });
+    return bidFloor.floor;
+  } catch (_) {
+    return 0
+  }
+}
+
+/**
+ * Create a new renderer
+ *
+ * @param bidRequest
+ * @param response
+ * @returns {Renderer}
+ */
+function newRenderer(bidRequest, response) {
+  logMessage('DSPx: newRenderer', bidRequest, response);
+  const renderer = Renderer.install({
+    id: response.renderer.id || response.bid_id,
+    url: (bidRequest.params && bidRequest.params.rendererUrl) || response.renderer.url,
+    config: response.renderer.options || deepAccess(bidRequest, 'renderer.options'),
+    loaded: false
+  });
+
+  try {
+    renderer.setRender(outstreamRender);
+  } catch (err) {
+    logWarn('Prebid Error calling setRender on renderer', err);
+  }
+  return renderer;
+}
+
+/**
+ * Outstream Render Function
+ *
+ * @param bid
+ */
+function outstreamRender(bid) {
+  logMessage('DSPx: outstreamRender bid:', bid);
+  const embedCode = createOutstreamEmbedCode(bid);
+  try {
+    const inIframe = getBidIdParameter('iframe', bid.renderer.config);
+    if (inIframe && window.document.getElementById(inIframe).nodeName === 'IFRAME') {
+      const iframe = window.document.getElementById(inIframe);
+      let framedoc = iframe.contentDocument || (iframe.contentWindow && iframe.contentWindow.document);
+      framedoc.body.appendChild(embedCode);
+      if (typeof window.dspxRender === 'function') {
+        window.dspxRender(bid);
+      } else {
+        logError('[dspx][renderer] Error: dspxRender function is not found');
+      }
+      return;
+    }
+
+    const slot = getBidIdParameter('slot', bid.renderer.config) || bid.adUnitCode;
+    if (slot && window.document.getElementById(slot)) {
+      window.document.getElementById(slot).appendChild(embedCode);
+      if (typeof window.dspxRender === 'function') {
+        window.dspxRender(bid);
+      } else {
+        logError('[dspx][renderer] Error: dspxRender function is not found');
+      }
+    } else if (slot) {
+      logError('[dspx][renderer] Error: slot not found');
+    }
+  } catch (err) {
+    logError('[dspx][renderer] Error:' + err.message)
+  }
+}
+
+/**
+ * create Outstream Embed Code Node
+ *
+ * @param bid
+ * @returns {DocumentFragment}
+ */
+function createOutstreamEmbedCode(bid) {
+  const fragment = window.document.createDocumentFragment();
+  let div = window.document.createElement('div');
+  div.innerHTML = bid.renderer.config.code;
```

Comment:
May be safer to use `deepAccess` to retrieve `code` from `bid.renderer.config` here

Reply:
thanks, improvement commit - done.